test/test_hash_picker.cpp

```diff
@@ -0,0 +1,548 @@
+	{
+		std::uint32_t hash[32 / 4];
+		std::fill(std::begin(hash), std::end(hash), i + 1);
+		full_tree[full_tree.end_index() - num_leafs + i] = sha256_hash(reinterpret_cast<char*>(hash));
+	}
+
+	merkle_fill_tree(full_tree, num_leafs);
+	return full_tree;
+}
+
+}
+
+#if 0
+TORRENT_TEST(pick_piece_layer)
+{
+	file_storage fs;
+	fs.set_piece_length(16 * 1024);
+
+	fs.add_file("test/tmp1", 4 * 512 * 16 * 1024);
+	fs.add_file("test/tmp2", 4 * 512 * 16 * 1024);
+
+	aux::vector<aux::vector<sha256_hash>, file_index_t> trees;
+
+	trees.push_back(aux::vector<sha256_hash>(merkle_num_nodes(merkle_num_leafs(4 * 512))));
+	aux::from_hex("0000000000000000000000000000000000000000000000000000000000000001", trees.back()[0].data());
+	trees.push_back(aux::vector<sha256_hash>(merkle_num_nodes(merkle_num_leafs(4 * 512))));
+	aux::from_hex("0000000000000000000000000000000000000000000000000000000000000001", trees.back()[0].data());
+
+	hash_picker picker(fs, trees);
+
+	typed_bitfield<piece_index_t> pieces;
+	pieces.resize(8 * 512);
+	pieces.set_all();
+
+	mock_peer_connection mock_peer1, mock_peer2;
+	mock_peer1.m_torrent_peer = (torrent_peer*)0x1;
+	mock_peer2.m_torrent_peer = (torrent_peer*)0x2;
+
+	auto picked = picker.pick_hashes(pieces, 2, &mock_peer1);
+	TEST_EQUAL(int(picked.size()), 2);
+	TEST_EQUAL(picked[0].file, 0_file);
+	TEST_EQUAL(picked[0].base, 0);
+	TEST_EQUAL(picked[0].count, 512);
+	TEST_EQUAL(picked[0].index, 0);
+	TEST_EQUAL(picked[0].proof_layers, 10);
+	TEST_EQUAL(picked[1].file, 0_file);
+	TEST_EQUAL(picked[1].base, 0);
+	TEST_EQUAL(picked[1].count, 512);
+	TEST_EQUAL(picked[1].index, 512);
+	TEST_EQUAL(picked[1].proof_layers, 10);
+
+	picked = picker.pick_hashes(pieces, 3, &mock_peer2);
+	TEST_EQUAL(int(picked.size()), 3);
+	TEST_EQUAL(picked[0].file, 0_file);
+	TEST_EQUAL(picked[0].base, 0);
+	TEST_EQUAL(picked[0].count, 512);
+	TEST_EQUAL(picked[0].index, 1024);
+	TEST_EQUAL(picked[0].proof_layers, 10);
+	TEST_EQUAL(picked[1].file, 0_file);
+	TEST_EQUAL(picked[1].base, 0);
+	TEST_EQUAL(picked[1].count, 512);
+	TEST_EQUAL(picked[1].index, 1536);
+	TEST_EQUAL(picked[1].proof_layers, 10);
+	TEST_EQUAL(picked[2].file, 1_file);
+	TEST_EQUAL(picked[2].base, 0);
+	TEST_EQUAL(picked[2].count, 512);
+	TEST_EQUAL(picked[2].index, 0);
+	TEST_EQUAL(picked[2].proof_layers, 10);
+
+	picked = picker.pick_hashes(pieces, 4, &mock_peer1);
+	TEST_EQUAL(int(picked.size()), 3);
+	TEST_EQUAL(picked[0].file, 1_file);
+	TEST_EQUAL(picked[0].base, 0);
+	TEST_EQUAL(picked[0].count, 512);
+	TEST_EQUAL(picked[0].index, 512);
+	TEST_EQUAL(picked[0].proof_layers, 10);
+	TEST_EQUAL(picked[1].file, 1_file);
+	TEST_EQUAL(picked[1].base, 0);
+	TEST_EQUAL(picked[1].count, 512);
+	TEST_EQUAL(picked[1].index, 1024);
+	TEST_EQUAL(picked[1].proof_layers, 10);
+	TEST_EQUAL(picked[2].file, 1_file);
+	TEST_EQUAL(picked[2].base, 0);
+	TEST_EQUAL(picked[2].count, 512);
+	TEST_EQUAL(picked[2].index, 1536);
+	TEST_EQUAL(picked[2].proof_layers, 10);
+}
+#endif
```

Comment:
is this test failing?

Reply:
It needs a re-write because the signature of `pick_hashes` changed and it now only returns one hash request at a time.